sinatra-contrib/lib/sinatra/namespace.rb

```diff
@@ -305,6 +305,13 @@ def pattern
         @pattern
       end
 
+      # because caller_locations is defined privately and method_missing calls
+      # send this will end up using ruby's caller_locations instead of
+      # calling base.method_missing back up to the sinatra definition.
+      def caller_locations
+        base.caller_locations
+      end
```

Comment:
This is wrong.  The `caller_locations` defined in sinatra should return array which is  cleaned up by `cleaned_caller`.
For example, your test shows wrong file location and lineno. Please take a look at the following:
```ruby
# got
["/Users/namusyaka/workspace/private/ruby/sinatra/sinatra-contrib/lib/sinatra/namespace.rb", "341"]

# expected
["/Users/namusyaka/workspace/private/ruby/sinatra/sinatra-contrib/test/namespace_spec.rb", "709"]
```

In order to fix the issue correctly, we should extract the original sinatra methods into module, and then include the module in `Sinatra` and `Sinatra::Namespace`